tests/dir_helpers.py

```diff
@@ -81,6 +73,28 @@ def __new__(cls, *args, **kwargs):
     def __fspath__(self):
         return str(self)
 
+    def init(self, *, scm=False, dvc=False):
```

Comment:
How about having two methods: `TmpDir::dvc_init()` and `TmpDir::git_init` that only injects `dvc` and `scm` to the instance?

Reply:
I tried, these two are interleaved too much, `.scm` construction depends on `.dvc` presense for example. Separate init methods make it less readable. You can look at current convoluted `scm` and `dvc` implementation.

Reply:
I might be missing something, but for, `TmpDir::dvc_init()`, wouldn't following work:
```python
self.dvc = Repo(path, no_scm=not hasattr(self, "scm"))
```
And, similarly, on `git_init()`: `self.scm = Git(str_path)`?

I just wanted it to be a bit flexible, but, without any magic, so that it'd be easier to just create a temp_dir and make anything out of it (such as git repo, dvc repo, or a normal directory).

Reply:
Only I don't want to instantiate `Git` twice. This caused issues under Windows as far as I remember. So you can't assign `scm` in `init_scm()`, you need to postpone that. It ended up like this:
```python
class TmpDir(...):
    def __new__(...):
        # ...
        self.dvc = self.scm = None
        # ...

    @cached_property
    def scm(self):
        from dvc.scm.git import Git

        return self.dvc.scm if self.dvc else Git(fspath(self))

    def init_scm(self):
        assert not self.scm

        _git_init(fspath(self))
        del self.scm  # Might be dvc.scm or a separate Git instance so we delay

    def init_dvc(self):
        from dvc.repo import Repo
        from dvc.scm import NoSCM

        assert not self.dvc

        self.dvc = Repo.init(fspath(self), no_scm=True)
        if not isinstance(self.dvc.scm, NoSCM):
            self.scm = self.dvc.scm
            self.scm.commit("init dvc")
```

As I said they are too interleaved.

Reply:
Got it, thanks for the explanation.